File: src/main/java/io/github/qingmo/json/internal/TypeUtils.java
/*
 * MIT License
 *
 * Copyright (c) 2021 qingmo(eagleqingluo@gmail.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package io.github.qingmo.json.internal;

import io.github.qingmo.json.exception.JSONException;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.RoundingMode;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class TypeUtils {
    private TypeUtils() {
        throw new UnsupportedOperationException("can not instance static class");
    }

    private final static Pattern NUMBER_WITH_TRAILING_ZEROS_PATTERN = Pattern.compile("\\.0*$");

    public static String castToString(Object value) {
        return value == null ? null : value.toString();
    }

    public static Byte castToByte(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof BigDecimal) {
            return byteValue((BigDecimal) value);
        }
        if (value instanceof Number) {
            return ((Number) value).byteValue();
        }
        if (value instanceof String) {
            String strVal = (String) value;
            if (strVal.length() == 0 //
                    || "null".equalsIgnoreCase(strVal)
            ) {
                return null;
            } else {
                return Byte.valueOf(strVal);
            }
        }
        if (value instanceof Boolean) {
            return (Boolean) value ? (byte) 1 : (byte) 0;
        }
        throw new JSONException("can not cast to byte, value : " + value);
    }

    public static Character castToChar(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof Character) {
            return (Character) value;
        }
        if (value instanceof String) {
            String strVal = (String) value;
            if (strVal.length() == 0) {
                return null;
            }
            if (strVal.length() != 1) {
                throw new JSONException("can not cast to char, value : " + value);
            }
            return strVal.charAt(0);
        }
        throw new JSONException("can not cast to char, value : " + value);
    }

    public static Short castToShort(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof BigDecimal) {
            return ((BigDecimal) value).shortValue();
        }
        if (value instanceof Number) {
            return ((Number) value).shortValue();
        }
        if (value instanceof String) {
            String strVal = (String) value;
            if (strVal.length() == 0 //
                    || "null".equalsIgnoreCase(strVal)
            ) {
                return null;
            } else {
                return Short.parseShort(strVal);
            }
        }
        if (value instanceof Boolean) {
            return (Boolean) value ? (short) 1 : (short) 0;
        }
        throw new JSONException("can not cast to short, value : " + value);
    }

    public static BigDecimal castToBigDecimal(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof Float) {
            Float ret = (Float) value;
            if (Float.isNaN(ret) || Float.isInfinite(ret)) {
                return null;
            }
        } else if (value instanceof Double) {
            Double ret = (Double) value;
            if (Double.isNaN(ret) || Double.isInfinite(ret)) {
                return null;
            }
        } else if (value instanceof BigDecimal) {
            return (BigDecimal) value;
        } else if (value instanceof BigInteger) {
            return new BigDecimal((BigInteger) value);
        }
        String strVal = value.toString();
        if (strVal.isEmpty()
                || "null".equalsIgnoreCase(strVal)
        ) {
            return null;
        }
        if (strVal.length() > 65535) {
            throw new JSONException("decimal overflow");
        }
        return new BigDecimal(strVal);
    }

    public static BigInteger castToBigInteger(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof Float) {
            Float floatValue = (Float) value;
            if (java.lang.Float.isNaN(floatValue) || java.lang.Float.isInfinite(floatValue)) {
                return null;
            } else {
                return BigInteger.valueOf(floatValue.longValue());
            }
        } else if (value instanceof Double) {
            Double doubleValue = (Double) value;
            if (java.lang.Double.isNaN(doubleValue) || java.lang.Double.isInfinite(doubleValue)) {
                return null;
            } else {
                return BigInteger.valueOf(doubleValue.longValue());
            }
        } else if (value instanceof BigInteger) {
            return (BigInteger) value;
        } else if (value instanceof BigDecimal) {
            return ((BigDecimal) value).toBigInteger();
        }
        String strVal = value.toString();
        if (strVal.isEmpty()
                || "null".equalsIgnoreCase(strVal)
        ) {
            return null;
        }
        if (strVal.length() > 65535) {
            throw new JSONException("decimal overflow");
        }
        return new BigInteger(strVal);
    }

    public static Float castToFloat(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof Number) {
            return ((Number) value).floatValue();
        }
        if (value instanceof String) {
            String strVal = (String) value;
            if (strVal.length() == 0 //
                    || "null".equalsIgnoreCase(strVal)
            ) {
                return null;
            }
            if (strVal.indexOf(',') != -1) {
                strVal = strVal.replace(",", "");
            }
            return Float.valueOf(strVal);
        }
        if (value instanceof Boolean) {
            return (Boolean) value ? 1f : 0f;
        }
        throw new JSONException("can not cast to float, value : " + value);
    }

    public static Double castToDouble(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof Number) {
            return ((Number) value).doubleValue();
        }
        if (value instanceof String) {
            String strVal = (String) value;
            if (strVal.length() == 0 //
                    || "null".equalsIgnoreCase(strVal)
            ) {
                return null;
            }
            if (strVal.indexOf(',') != -1) {
                strVal = strVal.replace(",", "");
            }
            return Double.valueOf(strVal);
        }
        if (value instanceof Boolean) {
            return (Boolean) value ? 1.0 : 0.0;
        }
        throw new JSONException("can not cast to double, value : " + value);
    }

    public static Long castToLong(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof BigDecimal) {
            return ((BigDecimal) value).longValue();
        }
        if (value instanceof Number) {
            return ((Number) value).longValue();
        }
        if (value instanceof String) {
            String strVal = (String) value;
            if (strVal.length() == 0 //
                    || "null".equalsIgnoreCase(strVal)
            ) {
                return null;
            }
            if (strVal.indexOf(',') != -1) {
                strVal = strVal.replace(",", "");
            }
            try {
                return Long.valueOf(strVal);
            } catch (NumberFormatException ex) {
                throw ex;
            }
        }
        if (value instanceof Boolean) {
            return (Boolean) value ? 1L : 0L;
        }
        throw new JSONException("can not cast to long, value : " + value);
    }

    public static Byte byteValue(BigDecimal decimal) {
        if (decimal == null) {
            return 0;
        }
        return decimal.setScale(0, RoundingMode.HALF_UP).byteValueExact();
    }

    public static Short shortValue(BigDecimal decimal) {
        if (decimal == null) {
            return 0;
        }
        return decimal.shortValueExact();
    }

    public static int intValue(BigDecimal decimal) {
        if (decimal == null) {
            return 0;
        }
        return decimal.intValueExact();
    }

    public static long longValue(BigDecimal decimal) {
        if (decimal == null) {
            return 0;
        }
        return decimal.longValueExact();
    }

    public static Integer castToInt(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof Integer) {
            return (Integer) value;
        }
        if (value instanceof BigDecimal) {
            return ((BigDecimal) value).intValue();
        }
        if (value instanceof Number) {
            return ((Number) value).intValue();
        }
        if (value instanceof String) {
            String strVal = (String) value;
            if (strVal.length() == 0
                    || "null".equalsIgnoreCase(strVal)
            ) {
                return null;
            }
            if (strVal.indexOf(',') != -1) {
                strVal = strVal.replace(",", "");
            }
            Matcher matcher = NUMBER_WITH_TRAILING_ZEROS_PATTERN.matcher(strVal);
            if (matcher.find()) {
                strVal = matcher.replaceAll("");
            }
            return Integer.valueOf(strVal);
        }
        if (value instanceof Boolean) {
            return ((Boolean) value) ? 1 : 0;
        }

        throw new JSONException("can not cast to int, value :" + value);
    }

    public static Boolean castToBoolean(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof Boolean) {
            return (Boolean) value;
        }
        if (value instanceof BigDecimal) {
            return ((BigDecimal) value).intValue() == 1;
        }
        if (value instanceof Number) {
            return ((Number) value).intValue() == 1;
        }
        if (value instanceof String) {
            String strVal = (String) value;
            if (strVal.length() == 0 //
                    || "null".equalsIgnoreCase(strVal)
            ) {
                return null;
            }
            if ("true".equalsIgnoreCase(strVal) //
                    || "1".equalsIgnoreCase(strVal)
            ) {
                return java.lang.Boolean.TRUE;
            }
            if ("false".equalsIgnoreCase(strVal) //
                    || "0".equalsIgnoreCase(strVal)
            ) {
                return java.lang.Boolean.FALSE;
            }
            if ("Y".equalsIgnoreCase(strVal) //
                    || "T".equalsIgnoreCase(strVal)
            ) {
                return java.lang.Boolean.TRUE;
            }
            if ("F".equalsIgnoreCase(strVal) //
                    || "N".equalsIgnoreCase(strVal)
            ) {
                return java.lang.Boolean.FALSE;
            }
        }
        throw new JSONException("can not cast to boolean, value :" + value);
    }
}

File: src/main/kotlin/io/github/qingmo/json/internal/TypeUtils.kt
/*
 * MIT License
 *
 * Copyright (c) 2021 qingmo(eagleqingluo@gmail.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package io.github.qingmo.json.internal

import io.github.qingmo.json.exception.JSONException
import java.math.BigDecimal
import java.math.BigInteger
import java.util.regex.Matcher
import java.util.regex.Pattern

object TypeUtils {
    private val NUMBER_WITH_TRAILING_ZEROS_PATTERN: Pattern = Pattern.compile("\\.0*$")
    fun castToString(value: Any?): String? {
        return value?.toString()
    }

    fun castToByte(value: Any?): Byte? {
        if (value == null) {
            return null
        }
        if (value is BigDecimal) {
            return byteValue(value as BigDecimal?)
        }
        if (value is Number) {
            return value.toByte()
        }
        if (value is String) {
            val strVal = value
            return if (strVal.length == 0 //
                || "null".equals(strVal, ignoreCase = true)
            ) {
                null
            } else strVal.toByte()
        }
        if (value is Boolean) {
            return if (value) 1.toByte() else 0.toByte()
        }
        throw JSONException("can not cast to byte, value : $value")
    }

    fun castToChar(value: Any?): Char? {
        if (value == null) {
            return null
        }
        if (value is Char) {
            return value
        }
        if (value is String) {
            val strVal = value
            if (strVal.length == 0) {
                return null
            }
            if (strVal.length != 1) {
                throw JSONException("can not cast to char, value : $value")
            }
            return strVal[0]
        }
        throw JSONException("can not cast to char, value : $value")
    }

    fun castToShort(value: Any?): Short? {
        if (value == null) {
            return null
        }
        if (value is BigDecimal) {
            return shortValue(value as BigDecimal?)
        }
        if (value is Number) {
            return value.toShort()
        }
        if (value is String) {
            val strVal = value
            return if (strVal.length == 0 //
                || "null".equals(strVal, ignoreCase = true)
            ) {
                null
            } else strVal.toShort()
        }
        if (value is Boolean) {
            return if (value) 1.toShort() else 0.toShort()
        }
        throw JSONException("can not cast to short, value : $value")
    }

    fun castToBigDecimal(value: Any?): BigDecimal? {
        if (value == null) {
            return null
        }
        if (value is Float) {
            if (java.lang.Float.isNaN((value as Float?)!!) || java.lang.Float.isInfinite((value as Float?)!!)) {
                return null
            }
        } else if (value is Double) {
            if (java.lang.Double.isNaN((value as Double?)!!) || java.lang.Double.isInfinite((value as Double?)!!)) {
                return null
            }
        } else if (value is BigDecimal) {
            return value
        } else if (value is BigInteger) {
            return BigDecimal(value as BigInteger?)
        } else if (value is Map<*, *> && value.size == 0) {
            return null
        }
        val strVal = value.toString()
        if (strVal.isEmpty()
            || "null".equals(strVal, ignoreCase = true)
        ) {
            return null
        }
        if (strVal.length > 65535) {
            throw JSONException("decimal overflow")
        }
        return BigDecimal(strVal)
    }

    fun castToBigInteger(value: Any?): BigInteger? {
        if (value == null) {
            return null
        }
        if (value is Float) {
            val floatValue = value
            return if (java.lang.Float.isNaN(floatValue) || java.lang.Float.isInfinite(floatValue)) {
                null
            } else BigInteger.valueOf(floatValue.toLong())
        } else if (value is Double) {
            val doubleValue = value
            return if (java.lang.Double.isNaN(doubleValue) || java.lang.Double.isInfinite(doubleValue)) {
                null
            } else BigInteger.valueOf(doubleValue.toLong())
        } else if (value is BigInteger) {
            return value
        } else if (value is BigDecimal) {
            return value.toBigInteger()
        }
        val strVal = value.toString()
        if (strVal.isEmpty()
            || "null".equals(strVal, ignoreCase = true)
        ) {
            return null
        }
        if (strVal.length > 65535) {
            throw JSONException("decimal overflow")
        }
        return BigInteger(strVal)
    }

    fun castToFloat(value: Any?): Float? {
        if (value == null) {
            return null
        }
        if (value is Number) {
            return value.toFloat()
        }
        if (value is String) {
            var strVal = value.toString()
            if (strVal.length == 0 //
                || "null".equals(strVal, ignoreCase = true)
            ) {
                return null
            }
            if (strVal.indexOf(',') != -1) {
                strVal = strVal.replace(",".toRegex(), "")
            }
            return strVal.toFloat()
        }
        if (value is Boolean) {
            return if (value) 1f else 0f
        }
        throw JSONException("can not cast to float, value : $value")
    }

    fun castToDouble(value: Any?): Double? {
        if (value == null) {
            return null
        }
        if (value is Number) {
            return value.toDouble()
        }
        if (value is String) {
            var strVal = value.toString()
            if (strVal.length == 0 //
                || "null".equals(strVal, ignoreCase = true)
            ) {
                return null
            }
            if (strVal.indexOf(',') != -1) {
                strVal = strVal.replace(",".toRegex(), "")
            }
            return strVal.toDouble()
        }
        if (value is Boolean) {
            return if (value) 1.0 else 0.0
        }
        throw JSONException("can not cast to double, value : $value")
    }

    fun castToLong(value: Any?): Long? {
        if (value == null) {
            return null
        }
        if (value is BigDecimal) {
            return longValue(value as BigDecimal?)
        }
        if (value is Number) {
            return value.toLong()
        }
        if (value is String) {
            var strVal = value
            if (strVal.length == 0 //
                || "null".equals(strVal, ignoreCase = true)
            ) {
                return null
            }
            if (strVal.indexOf(',') != -1) {
                strVal = strVal.replace(",".toRegex(), "")
            }
            return try {
                strVal.toLong()
            } catch (ex: NumberFormatException) {
                throw ex
            }
        }
        if (value is Boolean) {
            return if (value) 1L else 0L
        }
        throw JSONException("can not cast to long, value : $value")
    }

    fun byteValue(decimal: BigDecimal?): Byte {
        if (decimal == null) {
            return 0
        }
        return decimal.toBigInteger().toBigDecimal().byteValueExact()
    }

    fun shortValue(decimal: BigDecimal?): Short {
        if (decimal == null) {
            return 0
        }
        return decimal.toBigInteger().toBigDecimal().shortValueExact()
    }

    fun intValue(decimal: BigDecimal?): Int {
        if (decimal == null) {
            return 0
        }
        return decimal.toBigInteger().toBigDecimal().intValueExact()
    }

    fun longValue(decimal: BigDecimal?): Long {
        if (decimal == null) {
            return 0
        }
        return decimal.toBigInteger().toBigDecimal().longValueExact()
    }

    fun castToInt(value: Any?): Int? {
        if (value == null) {
            return null
        }
        if (value is Int) {
            return value
        }
        if (value is BigDecimal) {
            return intValue(value as BigDecimal?)
        }
        if (value is Number) {
            return value.toInt()
        }
        if (value is String) {
            var strVal = value
            if (strVal.length == 0 //
                || "null".equals(strVal, ignoreCase = true)
            ) {
                return null
            }
            if (strVal.indexOf(',') != -1) {
                strVal = strVal.replace(",".toRegex(), "")
            }
            val matcher: Matcher = NUMBER_WITH_TRAILING_ZEROS_PATTERN.matcher(strVal)
            if (matcher.find()) {
                strVal = matcher.replaceAll("")
            }
            return (strVal as String).toInt()
        }
        if (value is Boolean) {
            return if (value) 1 else 0
        }

        throw JSONException("can not cast to int, value : $value")
    }

    fun castToBoolean(value: Any?): Boolean? {
        if (value == null) {
            return null
        }
        if (value is Boolean) {
            return value
        }
        if (value is BigDecimal) {
            return intValue(value as BigDecimal?) == 1
        }
        if (value is Number) {
            return value.toInt() == 1
        }
        if (value is String) {
            val strVal = value
            if (strVal.length == 0 //
                || "null".equals(strVal, ignoreCase = true)
            ) {
                return null
            }
            if ("true".equals(strVal, ignoreCase = true) //
                || "1" == strVal
            ) {
                return java.lang.Boolean.TRUE
            }
            if ("false".equals(strVal, ignoreCase = true) //
                || "0" == strVal
            ) {
                return java.lang.Boolean.FALSE
            }
            if ("Y".equals(strVal, ignoreCase = true) //
                || "T" == strVal
            ) {
                return java.lang.Boolean.TRUE
            }
            if ("F".equals(strVal, ignoreCase = true) //
                || "N" == strVal
            ) {
                return java.lang.Boolean.FALSE
            }
        }
        throw JSONException("can not cast to boolean, value : $value")
    }
}
